### PR TITLE
RANS Laplace regularizer on far-field volume pressure predictions

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Volume pressure predictions don't satisfy any physical constraints — the model can predict spatially inconsistent pressure fields. In incompressible RANS, the continuity equation (div(u) = 0) couples the velocity field to pressure. While we're predicting pressure directly, we can enforce a **Laplacian regularizer** on the predicted volume pressure: ∇²p should be smooth in the far field. This is a soft physics constraint that penalizes overly noisy or discontinuous pressure predictions.

The hypothesis: a Laplace smoothness regularizer on far-field volume pressure predictions will reduce volume_pressure errors by encouraging physically consistent pressure fields.

## Instructions

Implement a **RANS-inspired Laplace regularizer** on volume pressure in `train.py`:

1. **Identify far-field volume points**: points with SDF > 1.0m (far from car surface) and |y| < 2m, |z| < 2m (not near tunnel walls).

2. **Estimate Laplacian** using finite differences or k-nearest neighbors:
   ```python
   # For each far-field volume point i, find k=8 nearest neighbors j
   # Laplacian estimate: L(p_i) ≈ mean(p_j) - p_i  (discrete Laplace-Beltrami)
   # Regularizer: ||L(p_pred)||^2 should be small
   
   knn_indices = ...  # precomputed or computed online
   p_neighbors = p_pred[knn_indices]  # (N, k)
   laplacian = p_neighbors.mean(dim=1) - p_pred  # (N,)
   laplace_loss = (laplacian[far_field_mask] ** 2).mean()
   
   total_loss = data_loss + lambda_laplace * laplace_loss
   ```

3. **Start with λ_laplace = 0.01** (weak regularizer), try λ = 0.1 if the first arm shows improvement.

4. Precompute the KNN graph at startup (not per-batch) for efficiency.

5. **Run with the full yi baseline config:**
```bash
cd target/
CUDA_VISIBLE_DEVICES=0,1,2,3 torchrun --nproc_per_node=4 train.py \
  --agent kohaku --wandb-group yi-round36-rans-laplace \
  --learnable-pe --optimizer lion --lr 1e-4 --weight-decay 5e-4 --clip-grad-norm 0.5 \
  --grad-ema-alpha 0.5 \
  --surface-curvature-features k1_k2 --beta-nll-beta 0.5 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --batch-size 4 --validation-every 1 --no-compile-model \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536
```

6. Log `train/laplace_loss` and `train/data_loss` separately to W&B so we can see the regularizer contribution.

**Decision at epoch 5:**
- Vol_pressure improved (< 12.0%): continue to completion.
- Laplace loss not converging (> 1e-2): reduce λ.
- Val_abupt > 9.5%: stop, reduce λ.

Report volume_pressure, surface_pressure, val_abupt, τ_y, τ_z per epoch. Include W&B run IDs.

## Baseline (yi, PR #576)

| Metric | yi best | AB-UPT target | Gap |
|--------|---------|---------------|-----|
| val_abupt | **8.2528%** | — | — |
| surface_pressure | 5.127% | 3.82% | 1.34× |
| wall_shear | 9.233% | 7.29% | 1.27× |
| volume_pressure | 12.438% | 6.08% | **2.05×** |
| τ_x | 7.815% | 5.35% | 1.46× |
| τ_y | 9.233% | 3.65% | **2.53×** |
| τ_z | 10.449% | 3.63% | **2.88×** |

Beat `val_abupt < 8.2528%` to land on yi. Primary target: volume_pressure < 12.0%.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: kohaku — please start the RANS Laplace regularizer experiment now. Your pods are running.

Quick-start checklist:
1. Precompute a KNN graph (k=8) over volume points at startup — do this once, not per-batch, to keep it efficient.
2. Define the far-field mask: volume points where distance from car body > some threshold (e.g., > 0.5 of bounding-box diagonal). Log how many points this covers.
3. Implement the Laplace loss: laplace_loss = mean((p_pred[i] - mean(p_pred[neighbors_of_i]))^2) over far-field points.
4. Start with lambda_laplace = 0.01. Log train/laplace_loss and train/data_loss separately to W&B.
5. Launch with the command in the PR body.
6. Post your W&B run ID here immediately.

EP1 watch: if laplace_loss is not decreasing alongside data_loss, your far-field mask may be too aggressive or too sparse — report the fraction of volume points in the far-field mask.

Decision at EP3 (budget gate — not EP5 as written, since we run ~2-3 epochs in the 6h window):
- If vol_pressure improved vs baseline (12.438%): report and continue.
- If val_abupt > 9.5% at EP3: reduce lambda to 0.001 for next arm.
- Primary target is vol_pressure < 12.0%.

**Comment by morganmcg1** [comment]
ADVISOR: kohaku — your pod has been running idle for over 6 days. I sent the full implementation checklist at 21:29 UTC yesterday. You have not posted a W&B run ID or responded.

**Please start the RANS Laplace regularizer experiment right now.** Do not wait.

Quick-launch command after implementing the regularizer:

```bash
cd /workspace/senpai/target
setsid nohup torchrun --standalone --nproc_per_node=4 train.py \
  --agent kohaku \
  --wandb-group yi-round36-laplace-reg \
  --wandb-name laplace-lambda1e-2-ffmask \
  --learnable-pe \
  --optimizer lion --lr 1e-4 --weight-decay 5e-4 --clip-grad-norm 0.5 \
  --grad-ema-alpha 0.5 \
  --beta-nll-beta 0.5 \
  --surface-curvature-features k1_k2 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --batch-size 4 --validation-every 1 --no-compile-model \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  > laplace_arm_a.log 2>&1 &
```

(The RANS Laplace regularizer is controlled by flags you are adding — include your new flags per the implementation spec.)

**Post W&B run ID immediately. EP1 gate: val_abupt ≤ 20% → continue; > 25% → something is wrong with the Laplace loss implementation.**

If you are stuck on the implementation, describe the specific blocker and I will help you resolve it. But do not let another hour pass without training running.

**Comment by morganmcg1** [comment]
ADVISOR: THIRD ESCALATION — your pod has been at 0% GPU for over 6 days despite two prior nudges. Pod logs show you were assigned PR #635 at iteration 316 (~20:07 UTC May 4) but no GPU activity has started since.

**The RANS Laplace regularizer must be started immediately.** You have all the information you need.

If there is a blocker (missing flag, import error, shape error), post the specific error message here and I will diagnose it. Do not spend more than 15 minutes debugging before posting here.

If the pod agent is stuck waiting for something, here is the minimal viable implementation:

**Minimal viable RANS Laplace loss (add to train.py or a helper):**


**Far-field mask (optional for first arm):** For Arm A, apply to ALL volume points first (simpler). If vol_pressure degrades, restrict to far-field (dist > 0.5m from surface).

**Start with:**


**Post W&B run ID within 5 minutes of launch.** EP1 gate: val_abupt ≤ 20% continue; > 25% post error log immediately.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #635 because ADVISOR: Closing after 3 escalations over 6+ days with zero GPU activity. Pod is READY but no training has started. The RANS Laplace regularizer hypothesis is valid and will be reassigned. Closing to free the student slot for a new experiment..

**Comment by morganmcg1** [comment]
STUDENT-WATCHDOG: stopping this student invocation on senpai-yi-kohaku-7788dd949b-5pp2m.

Reason: assignment changed from #635 to none for 1777951905s while train.py is active
Original assignment: #635
Current assignment: none

Active training processes before termination:

```
 348557  348555   25730 5311  0.2 python          python train.py --agent kohaku --wandb-name kohaku/laplace-smoke-test --wandb-group laplace-smoke --learnable-pe --optimizer lion --lr 1e-4 --weight-decay 5e-4 --clip-grad-norm 0.5 --grad-ema-alpha 0.5 --surface-curvature-features k1_k2 --beta-nll-beta 0.5 --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 --batch-size 1 --validation-every 99 --no-compile-model --epochs 1 --train-surface-points 4096 --eval-surface-points 4096 --train-volume-points 4096 --eval-volume-points 4096 --laplace-volume-weight 0.01 --laplace-k 8 --num-workers 0
```

The watchdog only takes this action after the assignment drift remains stable for 1800s, so transient GitHub/API wobbles should not trigger it.
